File: 2022/day13/day13.js
const input = require('./testdata/input.json')

function isValid(a, b) {
  if (typeof a === 'number' && typeof b === 'number') {
    if (a < b) {
      return 1
    }
    if (a === b) {
      return 0
    }
    return -1
  }

  if (Array.isArray(a) && typeof b === 'number') {
    b = [b]
  }

  if (Array.isArray(b) && typeof a === 'number') {
    a = [a]
  }

  if (Array.isArray(a) && Array.isArray(b)) {
    let i = 0
    while (i < a.length && i < b.length) {
      const x = isValid(a[i], b[i])
      if (x === 1) {
        return 1
      }
      if (x === -1) {
        return -1
      }
      i++
    }
    if (i == a.length) {
      if (a.length == b.length) {
        return 0
      }
      return 1
    }

    return -1
  }
}

function PartOne() {
  let sum = 0
  let pairNumber = 1
  for (let i = 0; i < input.length; i += 2) {
    const left = input[i]
    const right = input[i + 1]
    const validPair = isValid(left, right)
    if (validPair == 1) {
      sum += pairNumber
    }
    pairNumber++
  }
  console.log(sum)
}

function PartTwo() {
  input.push([[2]], [[6]])
  input.sort(isValid).reverse()
  let sum = 1
  for (let i = 0; i < input.length; i++) {
    if (input[i].length === 1 && input[i][0] == 2) {
      sum = sum * (i + 1)
    }

    if (input[i].length === 1 && input[i][0] == 6) {
      sum = sum * (i + 1)
    }
  }
  console.log(sum)
}

PartOne()
PartTwo()

File: 2022/day13/day13.go
package day13

import (
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"sort"
	"strings"
)

func cmp(a, b any) int {
	as, aok := a.([]any)
	bs, bok := b.([]any)

	switch {
	case !aok && !bok:
		return int(a.(float64) - b.(float64))
	case !aok:
		as = []any{a}
	case !bok:
		bs = []any{b}
	}

	for i := 0; i < len(as) && i < len(bs); i++ {
		if c := cmp(as[i], bs[i]); c != 0 {
			return c
		}
	}
	return len(as) - len(bs)
}

func parseInput(input io.Reader) ([]any, int) {
	rawInput, _ := ioutil.ReadAll(input)
	packages, total := []any{}, 0
	for i, s := range strings.Split(strings.TrimSpace(string(rawInput)), "\n\n") {
		s := strings.Split(s, "\n")
		var a, b any
		json.Unmarshal([]byte(s[0]), &a)
		json.Unmarshal([]byte(s[1]), &b)
		packages = append(packages, a, b)
		if cmp(a, b) <= 0 {
			total += i + 1
		}

	}
	return packages, total
}

func PartOne(input io.Reader) int {
	_, total := parseInput(input)
	return total
}

func PartTwo(input io.Reader) int {
	packages, _ := parseInput(input)
	packages = append(packages, []any{[]any{2.}}, []any{[]any{6.}})
	sort.Slice(packages, func(i, j int) bool { return cmp(packages[i], packages[j]) < 0 })
	total := 1
	for i, p := range packages {
		if fmt.Sprint(p) == "[[2]]" || fmt.Sprint(p) == "[[6]]" {
			total *= i + 1
		}
	}
	return total
}
